libp2p/switch.nim

```diff
@@ -407,41 +414,55 @@ proc mount*[T: LPProtocol](s: Switch, proto: T, matcher: Matcher = nil) {.gcsafe
 
   s.ms.addHandler(proto.codecs, proto, matcher)
 
-proc start*(s: Switch): Future[seq[Future[void]]] {.async, gcsafe.} =
-  trace "starting switch for peer", peerInfo = s.peerInfo
+proc accept(s: Switch, transport: Transport) {.async.} = # noraises
+  ## transport's accept loop
+  ##
 
-  proc handle(conn: Connection): Future[void] {.async, closure, gcsafe.} =
-    trace "Incoming connection", conn
+  while transport.running:
     try:
-      await s.upgradeIncoming(conn) # perform upgrade on incoming connection
+      trace "About to accept incoming connection"
+      let conn = await transport.accept()
+      trace "Accepted an incoming connection", conn
+      asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
+    except TransportClosedError as exc:
+      debug "Transport closed", exc = exc.msg
+      return
     except CancelledError as exc:
-      raise exc
+      trace "Canceling accept loop"
+      return
```

Comment:
In this case, this does not have a significant effect, however you introduced leak of `conn` on cancellation. So if you expect cancellation here you should close `conn` if it was created.